CompilerTemplate: partially rewrite loopy.ld

buhman opened the pull request:
- moved C-runtime initialization to crt0.c

- C-runtime initialization now includes C++ static initializers

- added C++ Makefile rule

- rom building is split to "elf" and "bin" parts; this makes it easier to view the outcome of running the linker script

- copy-pasted DWARF debugging sections from the default gcc linker script in a new "debug.ld" script keep things clean--without this, debugging information would not be included in the linker script output.

- enabled debugging information (line numbers, symbols, source code) for all compiler output

- none of the generated debugging information is included in the final binary--rom size requirements are unaffected

- generally rearranged and reformatted loopy.ld

- enabled -ffunction-sections, -fdata-sections, and --gc-sections. This has the effect that functions or data that are not referenced/reachable by the entry point (excluding KEEP() sections) are removed at link-time. This can be convenient for larger projects with somewhat lazily-maintained Makefiles.

- enabled link-time errors for unlinked sections. For example, this means that if one attempts to use C++ static destructors (what is the point in this case?), linking will fail with a message saying the static destructor input section is not placed by the linker script in any output section.

- improved the "stack reservation overlaps with static memory" concept in the linker script--this is now just subtracted from total RAM.

- enabled --print-memory-usage, to display static memory allocations.

I believe I preserved all of the behavior that was intended in the original linker script. I briefly tested this with a trivial example Loopy program, and compared the output of these changes byte-for-byte with a rom produced prior to these changes, and compared load-time static RAM allocation, and verified the stack pointer address is indeed at the end of RAM, but did not test farther than that.

buhman committed:
CompilerTemplate: partially rewrite loopy.ld

- moved C-runtime initialization to crt0.c

- C-runtime initialization now includes C++ static initializers

- added C++ Makefile rule

- rom building is split to "elf" and "bin" parts; this makes it easier
  to view the outcome of running the linker script

- copy-pasted DWARF debugging sections from the default gcc linker
  script in a new "debug.ld" script keep things clean--without this,
  debugging information would not be included in the linker script
  output.

- enabled debugging information (line numbers, symbols, source code)
  for all compiler output

- none of the generated debugging information is included in the final
  binary--rom size requirements are unaffected

- generally rearranged and reformatted loopy.ld

- enabled -ffunction-sections, -fdata-sections, and
  --gc-sections. This has the effect that functions or data that are not
  referenced/reachable by the entry point (excluding KEEP() sections)
  are removed at link-time. This can be convenient for larger projects
  with somewhat lazily-maintained Makefiles.

- enabled link-time errors for unlinked sections. For example, this
  means that if one attempts to use C++ static destructors (what is
  the point in this case?), linking will fail with a message saying
  the static destructor input section is not placed by the linker
  script in any output section.

- improved the "stack reservation overlaps with static memory" concept
  in the linker script--this is now just subtracted from total RAM.

- enabled --print-memory-usage, to display static memory allocations.

I believe I preserved all of the behavior that was intended in the
original linker script. I briefly tested this with a trivial example
Loopy program, and compared the output of these changes byte-for-byte
with a rom produced prior to these changes, and compared load-time
static RAM allocation, and verified the stack pointer address is
indeed at the end of RAM, but did not test farther than that.

buhman committed:
loopy.ld: adjust "last address" calculation

This uses the size of the data in rom as the "last address" in the
cartridge header.

- move gitignore directives to CompilerTemplate

- Makefile: remove LoopyMSE hack

- Makefile: use the ROM variable more consistently

- Makefile: move path separator inside the TOOLBIN variable

- make it possible to use a toolchain inside PATH.

kasamikona commented:
For some reason `__data_load_start` is getting the same address as `__data_link_start` (both in RAM) so the .data section is failing to load. I'm not sure what the correct fix for this is but this is why I didn't use `>RAM AT> ROM` originally. Any idea how to resolve?

buhman committed:
crt0.c: data_load_start typo

Previously this was making the C symbol `__data_load_start` equal to
the linker symbol `__data_link_start`, which was not intended.

buhman commented:
Sorry about that, [7fec9dc](https://github.com/kasamikona/Loopy-Tools/pull/2/commits/7fec9dcdabe7c23d703bcc282d02901f20f990a0)

kasamikona commented:
Confirmed working at least for C, the padding feature is missing but I'd like to rework that to just align to 4K pages (for current buggy LoopyMSE release). This is good for now.
